File: 2015/cs/AdventOfCode2015_CSharp/AdventOfCode2015.Core/Day02.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AdventOfCode2015.Core
{
    public class Day02 : Solution2015
    {
        public override int Day => 2;
        public override string Title => "I Was Told There Would Be No Math";

        public override string Run(string input)
        {
            List<Present> parsed = ParseInput(input);
            return FormatReport(SolvePartOne(parsed), SolvePartTwo(parsed));
        }

        public static int SolvePartOne(List<Present> presents)
        {
            return presents.Select(p => p.TotalPaper).Sum();
        }

        public static int SolvePartTwo(List<Present> presents) {
            return presents.Select(p => p.TotalRibbon).Sum();
        }

        public static List<Present> ParseInput(string input)
        {
            return input.Split('\n').Select(Present.FromString).ToList();
        }

        public struct Present
        {
            public readonly int Length;
            public readonly int Width;
            public readonly int Height;
            private readonly int[] ordered;

            public Present(int length, int width, int height)
            {
                this.Length = length;
                this.Width = width;
                this.Height = height;

                int[] dimensions = new int[] { length, width, height };
                Array.Sort(dimensions);
                ordered = dimensions;
            }

            public static Present FromString(string dimensions)
            {
                int[] parsed = dimensions.Split('x').Select(int.Parse).ToArray();
                if (parsed.Length != 3)
                {
                    throw new ArgumentException("Dimensions string should be three elements long (l, w, h).");
                }
                return new Present(length: parsed[0], width: parsed[1], height: parsed[2]);
            }

            // Part one
            public int TotalPaper => SurfaceArea + Slack;
            private int SurfaceArea => 2 * Length * Width + 2 * Width * Height + 2 * Height * Length;
            private int Slack => ordered[0] * ordered[1];

            // Part two
            public int TotalRibbon => 2 * ordered[0] + 2 * ordered[1] + Bow;
            private int Bow => ordered.Aggregate((total, dim) => total * dim);
        }

    }

}

File: 2015/java/src/main/java/com/robjwells/adventofcode2015/Day02.java
package com.robjwells.adventofcode2015;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Day02 extends Solution2015 {
    @Override
    public int getDay() {
        return 2;
    }

    @Override
    public String getTitle() {
        return "I Was Told There Would Be No Math";
    }

    @Override
    public String run(String input) {
        List<Present> parsed = parseInput(input);
        return formatReport(solvePartOne(parsed), solvePartTwo(parsed));
    }

    static class Present {
        private int length, width, height;
        private int[] ordered;

        Present(int length, int width, int height) {
            this.length = length;
            this.width = width;
            this.height = height;
            int[] dims = {length, width, height};
            Arrays.sort(dims);
            this.ordered = dims;
        }

        int surfaceArea() {
            return (2 * length * width +
                    2 * width * height +
                    2 * height * length
            );
        }

        int areaOfSmallestSide() {
            return ordered[0] * ordered[1];
        }

        int totalWrappingPaperNeeded() {
            return surfaceArea() + areaOfSmallestSide();
        }

        int smallestPerimeter() {
            return 2 * ordered[0] + 2 * ordered[1];
        }

        int volume() {
            return ordered[0] * ordered[1] * ordered[2];
        }

        int totalRibbonNeeded() {
            return smallestPerimeter() + volume();
        }

    }

    static Present parseInputLine(String line) {
        int[] dims = Arrays.stream(line.split("x")).mapToInt(Integer::parseInt).toArray();
        assert dims.length == 3;
        return new Present(dims[0], dims[1], dims[2]);
    }

    static List<Present> parseInput(String input) {
        return Arrays.stream(input.trim().split("\n"))
                .map(Day02::parseInputLine)
                .collect(Collectors.toList());
    }

    static int solvePartOne(List<Present> presents) {
        return presents.stream().mapToInt(Present::totalWrappingPaperNeeded).sum();
    }

    static int solvePartTwo(List<Present> presents) {
        return presents.stream().mapToInt(Present::totalRibbonNeeded).sum();
    }
}
